state/upgrades.go

```diff
@@ -3166,3 +3167,48 @@ func AddCharmOriginToApplications(pool *StatePool) error {
 		return nil
 	}))
 }
+
+// ExposeWildcardEndpointForExposedApplications adds an ExposedEndpoint entry
+// for the wildcard endpoint (to 0.0.0.0/0) for already exposed applications.
+// This ensures that all exposed applications are accessible at least one CIDR
+// and allows us to drop the fallback to 0.0.0.0/0 if no CIDRs present logic
+// from the firewaller worker.
+func ExposeWildcardEndpointForExposedApplications(pool *StatePool) error {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		col, closer := st.db().GetCollection(applicationsC)
+		defer closer()
+
+		var docs []applicationDoc
+		if err := col.Find(nil).All(&docs); err != nil {
+			return errors.Trace(err)
+		}
+
+		var implicitExposedEndpoints = map[string]ExposedEndpoint{
+			"": {
+				ExposeToCIDRs: []string{firewall.AllNetworksIPV4CIDR},
```

Comment:
I think it would be better to leave this for a future PR to make this feature-work IPV6-friendly. Mainly because I am unsure how the various firewall-supporting substrates would handle IPV6 CIDRs in firewall rules... 

I guess we would need quite a bit of testing to ensure that we don't break anything.

Reply:
Assuming IPV6 rules work for all substrates, it might be easier to add a check into the firewaller (right next to the if block introduced by this PR) and add the IPV6 wildcard if the set also contains the IPV4 one.

Reply:
@jameinel Any thoughts on ^^^?

Reply:
AWS says you need to explicitly list ::/0 along with 0.0.0.0/0
https://docs.aws.amazon.com/vpc/latest/userguide/vpc-migrate-ipv6.html

Openstack docs also seem to indicate the same:
https://docs.openstack.org/python-openstackclient/ussuri/cli/command-objects/security-group-rule.html

And we were already setting up ::/0
https://github.com/juju/juju/blob/0189bcb6e9fe202f84308c82adfc1e2c2f1addda/provider/openstack/firewaller.go#L377

Note that provider/ec2 was not:
https://github.com/juju/juju/blob/0189bcb6e9fe202f84308c82adfc1e2c2f1addda/provider/ec2/environ.go#L2049

However, I'm pretty sure that is because we have active deployments on Openstack that are using IPv6 but not on AWS (for now).

It is a bit more unclear on Azure and GCE, but it does seem to be the same.
It probably makes sense to do the conversion for all in the same upgrade step, so we don't have to worry about when/if someone ends up supporting one but not the other.